<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>The Bioconductor project: Working with annotations</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/08-annotations.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      The Bioconductor project
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            The Bioconductor project
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  The Bioconductor project
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 78%" class="percentage">
    78%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 78%" aria-valuenow="78" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/08-annotations.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-setup.html">1. Introduction and setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-introduction-to-bioconductor.html">2. Introduction to Bioconductor</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-installing-bioconductor.html">3. Installing Bioconductor packages</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-getting-help.html">4. Getting help</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-s4.html">5. S4 classes in Bioconductor</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-biological-sequences.html">6. Working with biological sequences</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-genomic-ranges.html">7. Working with genomics ranges</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        8. Working with annotations
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#install-packages">Install packages</a></li>
<li><a href="#overview">Overview</a></li>
<li><a href="#online-resources-or-bioconductor-annotation-packages">Online resources or Bioconductor annotation packages?</a></li>
<li><a href="#querying-annotations-from-ensembl-biomart">Querying annotations from Ensembl BioMart</a></li>
<li><a href="#querying-annotations-from-annotation-packages">Querying annotations from annotation packages</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-summarizedexperiment.html">9. The SummarizedExperiment class</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html"></a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="07-genomic-ranges.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="09-summarizedexperiment.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="07-genomic-ranges.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Working with
        </a>
        <a class="chapter-link float-end" href="09-summarizedexperiment.html" rel="next">
          Next: The...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Working with annotations</h1>
        <p>Last updated on 2025-06-09 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/08-annotations.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<hr><div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What Bioconductor packages provides methods to efficiently fetch and
use gene annotations?</li>
<li>How can I use gene annotation packages to convert between different
gene identifiers?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain how gene annotations are managed in the Bioconductor
project.</li>
<li>Identify Bioconductor packages and methods available to fetch and
use gene annotations.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a></h2>
<hr class="half-width"><p>Before we can proceed into the following sections, we install some
Bioconductor packages that we will need. First, we check that the <em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>
package is installed before trying to use it; otherwise we install it.
Then we use the <code>BiocManager::install()</code> function to install
the necessary packages.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"biomaRt"</span>, <span class="st">"org.Hs.eg.db"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a></h2>
<hr class="half-width"><p>Packages dedicated to query gene annotations exist in the ‘Software’
and ‘Annotation’ categories of the Bioconductor <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">biocViews</a>,
according to their nature.</p>
<p>In the ‘Software’ section, we find packages that do not actually
contain gene annotations, but rather dynamically <em>query</em> them
from online resources (e.g.,<a href="https://www.ensembl.org/biomart/martview" class="external-link">Ensembl BioMart</a>).
One such Bioconductor package is <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>.</p>
<p>Instead, in the ‘Annotation’ section, we find packages that do
contain annotations. Examples include <em><a href="https://bioconductor.org/packages/3.19/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>,
<em><a href="https://bioconductor.org/packages/3.19/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>,
and <em><a href="https://bioconductor.org/packages/3.19/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>.</p>
<p>In this episode, we will demonstrate the two approaches:</p>
<ul><li>Querying annotations from the Ensembl Biomart API using the <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
package.</li>
<li>Querying annotations from the <em><a href="https://bioconductor.org/packages/3.19/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
annotation package.</li>
</ul></section><section><h2 class="section-heading" id="online-resources-or-bioconductor-annotation-packages">Online resources or Bioconductor annotation packages?<a class="anchor" aria-label="anchor" href="#online-resources-or-bioconductor-annotation-packages"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="accessing-the-latest-information">Accessing the latest information<a class="anchor" aria-label="anchor" href="#accessing-the-latest-information"></a></h3>
<p>Bioconductor’s 6-month release cycle implies that packages available
from the latest stable release branch will not be updated for six months
(only bugfixes are allowed, not functional updates). As a result,
annotation packages may contain information that is out-of-date by up to
six months.</p>
<p>Instead, independent online resources often have different policies
driving the release of updated information. Some databases are
frequently updated, while others may not have been updated in years.</p>
<p>Accessing up-to-date information must also be balanced with
reproducibility. Having downloaded the ‘latest’ information at one point
is time is no good if one hasn’t recorded at least <em>when</em> that
information was downloaded.</p>
</div>
<div class="section level3">
<h3 id="storage-requirements">Storage requirements<a class="anchor" aria-label="anchor" href="#storage-requirements"></a></h3>
<p>By nature, Bioconductor annotation packages are larger than software
packages. Just like any other R package, annotation packages must be
installed on the user’s computer before they can be used. This can
rapidly use add up to using an amount of disk space that is not
negligible.</p>
<p>Conversely, online resources are generally accessed programmatically,
and generally only require users to record code to replicate analyses
reproducibly.</p>
</div>
<div class="section level3">
<h3 id="internet-connectivity">Internet connectivity<a class="anchor" aria-label="anchor" href="#internet-connectivity"></a></h3>
<p>When using online resources, it is often a good idea to write
annotations ownloaded from online resources to a local file, and refer
to that local file during analyses.</p>
<p>If online resources were to become unavailable for any reason (e.g.,
downtime, loss of internet connection), analyses that use local files
can carry on while those that rely on those online resources cannot.</p>
<p>In contrast, Bioconductor annotation packages only require internet
connectivity at the time of installation. Once installed, they do not
require internet connectivity, as they rely on information stored
locally.</p>
</div>
<div class="section level3">
<h3 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a></h3>
<p>Bioconductor annotation packages are naturally versioned, meaning
that users can confidently report the version of the package used in
their analysis. Just like software packages, users control if and when
annotation packages should be updated on their computer.</p>
<p>Online resources have different policies to facilitate reproducible
analyses. Some online resources keep archived versions of their
annotations, allowing users to consistently access the same information
over time. When this is not the case, it may be necessary to download a
copy of the annotation at one point in time, and preciously keep that
copy throughout the lifetime of the project to ensure the use of a
consistent set of annotations.</p>
</div>
<div class="section level3">
<h3 id="consistency">Consistency<a class="anchor" aria-label="anchor" href="#consistency"></a></h3>
<p>As we will see in the practical examples of this episode,
Bioconductor annotation packages generally re-use a consistent set of
data structures. This allows users familiar with one annotation package
to rapidly get started with others.</p>
<p>Independent online resources often organise their data in different
ways, which requires users to write custom code to access, retrieve, and
process their respective data.</p>
</div>
</section><section><h2 class="section-heading" id="querying-annotations-from-ensembl-biomart">Querying annotations from Ensembl BioMart<a class="anchor" aria-label="anchor" href="#querying-annotations-from-ensembl-biomart"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="the-ensembl-biomart">The Ensembl BioMart<a class="anchor" aria-label="anchor" href="#the-ensembl-biomart"></a></h3>
<p><a href="https://www.ensembl.org/biomart/martview" class="external-link">Ensembl
BioMart</a> is a robust data mining tool designed to facilitate access
to the vast array of biological data available through the Ensembl
project.</p>
<p>The <a href="https://www.ensembl.org/biomart/martview" class="external-link">BioMart web
interface</a> enables researchers to efficiently query and retrieve data
on genes, proteins, and other genomic features across multiple species.
It allows users to filter, sort, and export data based on various
attributes such as gene IDs, chromosomal locations, and functional
annotations.</p>
</div>
<div class="section level3">
<h3 id="the-bioconductor-biomart-package">The Bioconductor <code>biomaRt</code> package<a class="anchor" aria-label="anchor" href="#the-bioconductor-biomart-package"></a></h3>
<p><em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
is a Bioconductor software package that enables retrieval of large
amounts of data from Ensembl BioMart tables directly from an R session
where those annotations can be used.</p>
<p>Let us first load the package:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-available-marts">Listing available marts<a class="anchor" aria-label="anchor" href="#listing-available-marts"></a></h3>
<p>Ensembl BioMart organises its diverse biological information into
four databases also known as ‘marts’ or ‘biomarts’. Each mart focuses on
a different type of data.</p>
<p>Users must select the mart corresponds to the type of data they are
interested in before they can query any information from it.</p>
<p>The function <code>listMarts()</code> can be used to display the
names of those marts. This is convenient as users do not need to
memorise the name of the marts, and the function will also return an
updated list of names if any mart is renamed, added, or removed.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">listMarts</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               biomart                version
1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 114
2   ENSEMBL_MART_MOUSE      Mouse strains 114
3     ENSEMBL_MART_SNP  Ensembl Variation 114
4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 114</code></pre>
</div>
<p>In this demonstration, we will use the biomart called
<code>ENSEMBL_MART_ENSEMBL</code>, which contains the Ensembl gene
set.</p>
<p>Notably, the <code>version</code> columns also indicates the version
of the biomart. The Ensembl BioMart is updated regularly (multiple times
per year). By default, <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
functions access the latest version of each biomart. This is not ideal
for reproducibility.</p>
<p>Thankfully, Ensembl BioMart archives past versions of its mars in a
way that is accessible both programmatically, and on its website.</p>
<p>The function <code>listEnsemblArchives()</code> can be used to
display all the versions of Ensembl Biomart accessible.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">listEnsemblArchives</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             name     date                                 url version
1  Ensembl GRCh37 Feb 2014          https://grch37.ensembl.org  GRCh37
2     Ensembl 114 May 2025 https://may2025.archive.ensembl.org     114
3     Ensembl 113 Oct 2024 https://oct2024.archive.ensembl.org     113
4     Ensembl 112 May 2024 https://may2024.archive.ensembl.org     112
5     Ensembl 111 Jan 2024 https://jan2024.archive.ensembl.org     111
6     Ensembl 110 Jul 2023 https://jul2023.archive.ensembl.org     110
7     Ensembl 109 Feb 2023 https://feb2023.archive.ensembl.org     109
8     Ensembl 108 Oct 2022 https://oct2022.archive.ensembl.org     108
9     Ensembl 107 Jul 2022 https://jul2022.archive.ensembl.org     107
10    Ensembl 106 Apr 2022 https://apr2022.archive.ensembl.org     106
11    Ensembl 105 Dec 2021 https://dec2021.archive.ensembl.org     105
12    Ensembl 104 May 2021 https://may2021.archive.ensembl.org     104
13    Ensembl 103 Feb 2021 https://feb2021.archive.ensembl.org     103
14    Ensembl 102 Nov 2020 https://nov2020.archive.ensembl.org     102
15    Ensembl 101 Aug 2020 https://aug2020.archive.ensembl.org     101
16    Ensembl 100 Apr 2020 https://apr2020.archive.ensembl.org     100
17     Ensembl 99 Jan 2020 https://jan2020.archive.ensembl.org      99
18     Ensembl 80 May 2015 https://may2015.archive.ensembl.org      80
19     Ensembl 77 Oct 2014 https://oct2014.archive.ensembl.org      77
20     Ensembl 75 Feb 2014 https://feb2014.archive.ensembl.org      75
21     Ensembl 54 May 2009 https://may2009.archive.ensembl.org      54
   current_release
1
2                *
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21                </code></pre>
</div>
<p>In the output above, the key piece of information is the
<code>url</code> column, which provides the URL that <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
functions will need to access data from the corresponding snapshot of
the Ensembl BioMart.</p>
<p>At the time of writing, the current release is Ensembl 112, so let us
use the corresponding url
<code>https://may2024.archive.ensembl.org</code> to ensure reproducible
results no matter when this lesson is delivered.</p>
</div>
<div class="section level3">
<h3 id="connecting-to-a-biomart">Connecting to a biomart<a class="anchor" aria-label="anchor" href="#connecting-to-a-biomart"></a></h3>
<p>The two pieces of information collected above – the name of a biomart
and the URL of a snapshot – is all that is needed to connect to a
BioMart database reproducibly.</p>
<p>The function <code>useMart()</code> can then be used to create a
connection. The connection is traditionally stored in an object called
<code>mart</code>, to be reused in subsequent steps for querying
information from the online mart.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu">useMart</span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>, host <span class="op">=</span> <span class="st">"https://may2024.archive.ensembl.org"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-available-data-sets">Listing available data sets<a class="anchor" aria-label="anchor" href="#listing-available-data-sets"></a></h3>
<p>Each biomart contains a number of data sets.</p>
<p>The function <code>listDatasets()</code> can be used to display
information about those data sets. This is convenient as users do not
need to memorise the name of the data sets, and the information returned
by the function includes a short description of each data set, as well
as its version.</p>
<p>In the example below, we restrict the output table to the first few
rows, as the full table comprises 214 rows.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">listDatasets</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                       dataset                           description
1 abrachyrhynchus_gene_ensembl Pink-footed goose genes (ASM259213v1)
2     acalliptera_gene_ensembl      Eastern happy genes (fAstCal1.3)
3   acarolinensis_gene_ensembl       Green anole genes (AnoCar2.0v2)
4    acchrysaetos_gene_ensembl       Golden eagle genes (bAquChr1.2)
5    acitrinellus_gene_ensembl        Midas cichlid genes (Midas_v5)
6    amelanoleuca_gene_ensembl       Giant panda genes (ASM200744v2)
      version
1 ASM259213v1
2  fAstCal1.3
3 AnoCar2.0v2
4  bAquChr1.2
5    Midas_v5
6 ASM200744v2</code></pre>
</div>
<p>In the output above, the key piece of information is the
<code>dataset</code> column, which provides the identifier that <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
functions will need to access data from the corresponding biomart
table.</p>
<p>In this demonstration, we will use the Ensembl gene set for Homo
sapiens, which is not visible in the output above.</p>
<p>Given the number of data sets available, let us programmatically
filter the table of information using pattern matching rather than
searching the table manually:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subset</span><span class="op">(</span><span class="fu">listDatasets</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span>, <span class="fu">grepl</span><span class="op">(</span><span class="st">"sapiens"</span>, <span class="va">dataset</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                 dataset              description    version
80 hsapiens_gene_ensembl Human genes (GRCh38.p14) GRCh38.p14</code></pre>
</div>
<p>From the output above, we identify the desired data set identifier as
<code>hsapiens_gene_ensembl</code>.</p>
</div>
<div class="section level3">
<h3 id="connecting-to-a-data-set">Connecting to a data set<a class="anchor" aria-label="anchor" href="#connecting-to-a-data-set"></a></h3>
<p>Having chosen the data set that we want to use, we need to call the
function <code>useMart()</code> again, this time specifying the selected
data set.</p>
<p>Typically, one would copy paste the previous call to
<code>useMart()</code> and edit as needed. It is also common practice to
replace the <code>mart</code> object with the new connection.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu">useMart</span><span class="op">(</span></span>
<span>  biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>,</span>
<span>  dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span>,</span>
<span>  host <span class="op">=</span> <span class="st">"https://may2024.archive.ensembl.org"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-information-available-in-a-data-set">Listing information available in a data set<a class="anchor" aria-label="anchor" href="#listing-information-available-in-a-data-set"></a></h3>
<p>BioMart tables contain many pieces of information also known as
‘attributes’. So many, in fact, that they have been grouped into
categories also known as ‘pages’.</p>
<p>The function <code>listAttributes()</code> can be used to display
information about those attributes. This is convenient as users do not
need to memorise the name of the attributes, and the information
returned by the function includes a short description of each attribute,
as well as its page categorisation.</p>
<p>In the example below, we restrict the output table to the first few
rows, as the full table comprises 3157 rows.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">listAttributes</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                           name                  description         page
1               ensembl_gene_id               Gene stable ID feature_page
2       ensembl_gene_id_version       Gene stable ID version feature_page
3         ensembl_transcript_id         Transcript stable ID feature_page
4 ensembl_transcript_id_version Transcript stable ID version feature_page
5            ensembl_peptide_id            Protein stable ID feature_page
6    ensembl_peptide_id_version    Protein stable ID version feature_page</code></pre>
</div>
<p>In the output above, the key piece of information is the
<code>name</code> column, which provides the identifier that <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
functions will need to query that information from the corresponding
biomart data set.</p>
<p>The choice of attributes to query now depends on what it is we wish
to achieve.</p>
<p>For instance, let us imagine that we have a set of gene identifiers,
for which we wish to query:</p>
<ul><li>The gene symbol</li>
<li>The name of the chromosome where the gene is located</li>
<li>The start and end position of the gene on that chromosome</li>
<li>The strand on which the gene is encoded</li>
</ul><p>Users would often manually explore the full table of attributes to
identify the ones they wish to include in their query. It is also
possible to programmatically filter the table of attribute, based on
experience and intuition, to narrow down the search:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subset</span><span class="op">(</span><span class="fu">listAttributes</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span>, <span class="fu">grepl</span><span class="op">(</span><span class="st">"position"</span>, <span class="va">name</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu">grepl</span><span class="op">(</span><span class="st">"feature"</span>, <span class="va">page</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             name     description         page
10 start_position Gene start (bp) feature_page
11   end_position   Gene end (bp) feature_page</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="querying-information-from-a-biomart-table">Querying information from a BioMart table<a class="anchor" aria-label="anchor" href="#querying-information-from-a-biomart-table"></a></h3>
<p>We have now all the information that we need to perform the actual
query:</p>
<ul><li>A connection to a BioMart data set</li>
<li>The list of attributes available in that data set</li>
</ul><p>The function <code>getBM()</code> is the main <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
query function. Given a set of filters and corresponding values, it
retrieves the attributes requested by the user from the BioMart data set
it is connected to.</p>
<p>In the example below, we manually create a vector of arbitrary gene
identifiers for our query. In practice, the query will often originate
from an earlier analysis (e.g., differential gene expression).</p>
<p>The example below also queries attributes that we have not introduced
yet. In the previous section, we described how one may search the table
of attributes returned by <code>listAttributes()</code> to identify
attributes to include in their query.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_gene_ids</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span></span>
<span>  <span class="st">"ENSG00000133101"</span>,</span>
<span>  <span class="st">"ENSG00000145386"</span>,</span>
<span>  <span class="st">"ENSG00000134057"</span>,</span>
<span>  <span class="st">"ENSG00000157456"</span>,</span>
<span>  <span class="st">"ENSG00000147082"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">getBM</span><span class="op">(</span></span>
<span>  attributes <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"ensembl_gene_id"</span>,</span>
<span>    <span class="st">"hgnc_symbol"</span>,</span>
<span>    <span class="st">"chromosome_name"</span>,</span>
<span>    <span class="st">"start_position"</span>,</span>
<span>    <span class="st">"end_position"</span>,</span>
<span>    <span class="st">"strand"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>  values <span class="op">=</span> <span class="va">query_gene_ids</span>,</span>
<span>  mart <span class="op">=</span> <span class="va">mart</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in .processResults(postRes, mart = mart, hostURLsep = sep, fullXmlQuery = fullXmlQuery, : Query ERROR: caught BioMart::Exception::Database: Error during query execution: Table 'ensembl_mart_112.hsapiens_gene_ensembl__ox_hgnc__dm' doesn't exist</code></pre>
</div>
<p>Note that we also included the filtering attribute
<code>ensembl_gene_id</code> to the attributes retrieved from the data
set. This is key to reliably match the newly retrieved attributes to
those used in the query.</p>
</div>
</section><section><h2 class="section-heading" id="querying-annotations-from-annotation-packages">Querying annotations from annotation packages<a class="anchor" aria-label="anchor" href="#querying-annotations-from-annotation-packages"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="families-of-annotation-packages">Families of annotation packages<a class="anchor" aria-label="anchor" href="#families-of-annotation-packages"></a></h3>
<p>To balance the need for comprehensive information while maintaining
reasonable package sizes, Bioconductor annotation packages are organised
by release, data type, and species.</p>
<p>The major families of Bioconductor annotation packages are:</p>
<ul><li>
<code>OrgDb</code> packages provide mapping between various types of
gene identifiers and pathway information.</li>
<li>
<code>EnsDb</code> packages provide individual releases of Ensembl
annotations.</li>
<li>
<code>TxDb</code> packages provide individual releases of UCSC
annotations.</li>
</ul><p>All those families of annotations derive from the
<code>AnnotationDb</code> base class defined in the <em><a href="https://bioconductor.org/packages/3.19/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
package. As a result, any of those annotation packages can be accessed
using the same set of R functions, as demonstrated in the following
sections.</p>
</div>
<div class="section level3">
<h3 id="using-an-orgdb-package">Using an OrgDb package<a class="anchor" aria-label="anchor" href="#using-an-orgdb-package"></a></h3>
<p>In this example, we will use the <em><a href="https://bioconductor.org/packages/3.19/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
package to demonstrate the use of gene annotations for the human
species.</p>
<p>Let us first load the package:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<p>Each <code>OrgDb</code> package contains an object named identically
to the package itself. That object contains the annotations that the
package is meant to disseminate.</p>
<p>Aside from querying information, the whole object can be called to
print information about the annotations it contains, including the date
at which the snapshots of annotations that it contains were made.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">org.Hs.eg.db</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>OrgDb object:
| DBSCHEMAVERSION: 2.1
| Db type: OrgDb
| Supporting package: AnnotationDbi
| DBSCHEMA: HUMAN_DB
| ORGANISM: Homo sapiens
| SPECIES: Human
| EGSOURCEDATE: 2024-Mar12
| EGSOURCENAME: Entrez Gene
| EGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| CENTRALID: EG
| TAXID: 9606
| GOSOURCENAME: Gene Ontology
| GOSOURCEURL: http://current.geneontology.org/ontology/go-basic.obo
| GOSOURCEDATE: 2024-01-17
| GOEGSOURCEDATE: 2024-Mar12
| GOEGSOURCENAME: Entrez Gene
| GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| KEGGSOURCENAME: KEGG GENOME
| KEGGSOURCEURL: ftp://ftp.genome.jp/pub/kegg/genomes
| KEGGSOURCEDATE: 2011-Mar15
| GPSOURCENAME: UCSC Genome Bioinformatics (Homo sapiens)
| GPSOURCEURL:
| GPSOURCEDATE: 2024-Feb29
| ENSOURCEDATE: 2023-Nov22
| ENSOURCENAME: Ensembl
| ENSOURCEURL: ftp://ftp.ensembl.org/pub/current_fasta
| UPSOURCENAME: Uniprot
| UPSOURCEURL: http://www.UniProt.org/
| UPSOURCEDATE: Thu Apr 18 21:39:39 2024</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Please see: help('select') for usage information</code></pre>
</div>
<p>That same object is the one that needs to be supplied to <em><a href="https://bioconductor.org/packages/3.19/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
functions for running queries and retrieving annotations.</p>
</div>
<div class="section level3">
<h3 id="listing-information-available-in-an-annotation-package">Listing information available in an annotation package<a class="anchor" aria-label="anchor" href="#listing-information-available-in-an-annotation-package"></a></h3>
<p>The function <code>columns()</code> can be used to display the
annotations available in the object.</p>
<p>Here, the word ‘column’ refers to columns of tables used to store
information in database, the very same concept as ‘attributes’ in
BioMart. In other words, columns represent all the types of annotations
that may be retrieved from the object.</p>
<p>This is convenient as users do not need to memorise the names of the
columns of annotations available in the package.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">columns</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"
[26] "UNIPROT"     </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-keys-and-key-types">Listing keys and key types<a class="anchor" aria-label="anchor" href="#listing-keys-and-key-types"></a></h3>
<p>In database terminology, <em>keys</em> are the values by which
information may be queried from a database table.</p>
<p>Information being organised in columns, <em>key types</em> are the
names of the columns in which the key values are stored.</p>
<p>Given the variable number of columns in database tables, some tables
may allow information to be queried by more than one key. As a result,
it is crucial to specify both the keys and the type of key as part of
the query.</p>
<p>The function <code>keytypes()</code> can be used to display the names
of the columns that may be used to query information from the
object.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">keytypes</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"
[26] "UNIPROT"     </code></pre>
</div>
<p>The function <code>keys()</code> can be used to display all the
possible values for a given key type.</p>
<p>It is generally better practice to specify the type of key being
queried (to avoid ambiguity), although database tables typically have a
‘primary key’ used if users do not specify a type themselves.</p>
<p>In the example below, we restrict the list of gene symbol keys to the
first few values, as the full set comprises 193279 values.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">keys</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A1BG"  "A2M"   "A2MP1" "NAT1"  "NAT2"  "NATP" </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="querying-information-from-an-annotation-package">Querying information from an annotation package<a class="anchor" aria-label="anchor" href="#querying-information-from-an-annotation-package"></a></h3>
<p>The function <code>select()</code> is the main <em><a href="https://bioconductor.org/packages/3.19/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
query function. Given an <code>AnnotationDb</code> object, key values,
and columns (and optionally the type of key supplied if not the primary
key), it retrieves the columns requested by the user from the annotation
object.</p>
<p>In the example below, we re-use the vector of arbitrary gene
identifiers used in the BioMart example a few sections above.</p>
<p>As you can see from the output of the <code>columns()</code>
function, the annotation object does not contain some of the attributes
that we queried in the Biomart example. In this case, let us query:</p>
<ul><li>the gene symbol</li>
<li>the gene name</li>
<li>the gene type</li>
</ul><div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keys <span class="op">=</span> <span class="va">query_gene_ids</span>,</span>
<span>  columns <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"SYMBOL"</span>,</span>
<span>    <span class="st">"GENENAME"</span>,</span>
<span>    <span class="st">"GENETYPE"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>          ENSEMBL SYMBOL  GENENAME       GENETYPE
1 ENSG00000133101  CCNA1 cyclin A1 protein-coding
2 ENSG00000145386  CCNA2 cyclin A2 protein-coding
3 ENSG00000134057  CCNB1 cyclin B1 protein-coding
4 ENSG00000157456  CCNB2 cyclin B2 protein-coding
5 ENSG00000147082  CCNB3 cyclin B3 protein-coding</code></pre>
</div>
<p>One small but notable difference with <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
is that the output of <code>select()</code> automatically contains the
column that correspond to the key type used in the query. In other
words, there is no need to specify the key type(s) again in the
column(s) to retrieve.</p>
</div>
<div class="section level3">
<h3 id="vectorized-11-mapping">Vectorized 1:1 mapping<a class="anchor" aria-label="anchor" href="#vectorized-11-mapping"></a></h3>
<p>It is sometimes possible for annotations to display 1-to-many
relationships. For instance, individual genes typically have a unique
Ensembl gene identifier, while they may be known under multiple gene
name aliases.</p>
<p>The <code>select()</code> function demonstrated in the previous
section automatically returns <em>all</em> values in the columns
requested, for the key specified. This is possible thanks to the tabular
format in which annotations are returned; rows are added, repeating
values as necessary to display them on the same row as every other
values they are associated with.</p>
<p>In some cases, that behaviour is not desirable. Instead, users may
wish to retrieve a single value for each key that they input. One common
scenario arises during differential gene expression (DGE), where gene
identifiers are used to uniquely identify genes throughout the analysis,
while gene symbols are added to the final table of DGE statistics, to
provide more readable human-friendly gene identifiers. However, it is
not desirable to duplicate rows of DGE statistics, and thus only a
single gene symbol is required to annotate each gene.</p>
<p>The function <code>mapIds()</code> can be used for this purpose. A
major difference between the functions <code>mapIds()</code> and
<code>select()</code> are their arguments <code>column</code> (singular)
and <code>columns</code> (plural), respectively. The function
<code>mapIds()</code> accepts a single column name and returns a named
character vector where names are the input query values, and values are
the corresponding values in the requested column.</p>
<p>To deal with 1-to-many relationships, the function
<code>mapIds()</code> has an argument <code>multiVals</code> which can
be used to specify how the function should handle multiple values. The
default is to take the first value and ignore any other value.</p>
<p>In the example below, we query the gene symbol for a set of Ensembl
gene identifiers.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mapIds</span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keys <span class="op">=</span> <span class="va">query_gene_ids</span>,</span>
<span>  column <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ENSG00000133101 ENSG00000145386 ENSG00000134057 ENSG00000157456 ENSG00000147082
        "CCNA1"         "CCNA2"         "CCNB1"         "CCNB2"         "CCNB3" </code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Load the packages <em><a href="https://bioconductor.org/packages/3.19/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>
and <em><a href="https://bioconductor.org/packages/3.19/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>.
Then, display the columns of annotations available in those
packages.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="fu">columns</span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ENTREZID"            "EXONID"              "EXONIDX"
 [4] "EXONSEQEND"          "EXONSEQSTART"        "GENEBIOTYPE"
 [7] "GENEID"              "GENENAME"            "GENESEQEND"
[10] "GENESEQSTART"        "INTERPROACCESSION"   "ISCIRCULAR"
[13] "PROTDOMEND"          "PROTDOMSTART"        "PROTEINDOMAINID"
[16] "PROTEINDOMAINSOURCE" "PROTEINID"           "PROTEINSEQUENCE"
[19] "SEQCOORDSYSTEM"      "SEQLENGTH"           "SEQNAME"
[22] "SEQSTRAND"           "SYMBOL"              "TXBIOTYPE"
[25] "TXCDSSEQEND"         "TXCDSSEQSTART"       "TXID"
[28] "TXNAME"              "TXSEQEND"            "TXSEQSTART"
[31] "UNIPROTDB"           "UNIPROTID"           "UNIPROTMAPPINGTYPE" </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span>
<span><span class="fu">columns</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "CDSCHROM"   "CDSEND"     "CDSID"      "CDSNAME"    "CDSPHASE"
 [6] "CDSSTART"   "CDSSTRAND"  "EXONCHROM"  "EXONEND"    "EXONID"
[11] "EXONNAME"   "EXONRANK"   "EXONSTART"  "EXONSTRAND" "GENEID"
[16] "TXCHROM"    "TXEND"      "TXID"       "TXNAME"     "TXSTART"
[21] "TXSTRAND"   "TXTYPE"    </code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Bioconductor provides a wide range annotation packages.</li>
<li>Some Bioconductor software packages can be used to programmatically
access online resources.</li>
<li>Users should carefully choose their source of annotations based on
their needs and expectations.</li>
</ul></div>
</div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="07-genomic-ranges.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="09-summarizedexperiment.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="07-genomic-ranges.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Working with
        </a>
        <a class="chapter-link float-end" href="09-summarizedexperiment.html" rel="next">
          Next: The...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/08-annotations.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-project/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-project/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-project/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/bioc-project/08-annotations.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Working with annotations",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-project/08-annotations.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-project/08-annotations.html",
  "dateCreated": "2020-09-14",
  "dateModified": "2025-06-09",
  "datePublished": "2025-06-10"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

