<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>The Bioconductor project: Working with genomics ranges</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../07-genomic-ranges.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      The Bioconductor project
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            The Bioconductor project
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  The Bioconductor project
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 67%" class="percentage">
    67%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 67%" aria-valuenow="67" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../07-genomic-ranges.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-setup.html">1. Introduction and setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-introduction-to-bioconductor.html">2. Introduction to Bioconductor</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-installing-bioconductor.html">3. Installing Bioconductor packages</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-getting-help.html">4. Getting help</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-s4.html">5. S4 classes in Bioconductor</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-biological-sequences.html">6. Working with biological sequences</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        7. Working with genomics ranges
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#install-packages">Install packages</a></li>
<li><a href="#the-genomicranges-package-and-classes">The GenomicRanges package and classes</a></li>
<li><a href="#the-genomicranges-package">The GenomicRanges package</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-annotations.html">8. Working with annotations</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-summarizedexperiment.html">9. The SummarizedExperiment class</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/06-biological-sequences.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/08-annotations.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/06-biological-sequences.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Working with
        </a>
        <a class="chapter-link float-end" href="../instructor/08-annotations.html" rel="next">
          Next: Working with...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Working with genomics ranges</h1>
        <p>Last updated on 2025-06-09 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/07-genomic-ranges.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What is the recommended way to represent coordinates on a genomic
scale in Bioconductor?</li>
<li>What Bioconductor packages provides methods to efficiently process
genomic ranges?</li>
<li>How can I import/export sets of genomic coordinates from/to various
genomic file formats?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain how genomic coordinates and intervals are represented in the
Bioconductor project.</li>
<li>Identify Bioconductor packages and methods available to process
ranges of genomic coordinates.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a></h2>
<hr class="half-width"><p>Before we can proceed into the following sections, we install some
Bioconductor packages that we will need. First, we check that the <em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>
package is installed before trying to use it; otherwise we install it.
Then we use the <code>BiocManager::install()</code> function to install
the necessary packages.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"GenomicRanges"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="the-genomicranges-package-and-classes">The GenomicRanges package and classes<a class="anchor" aria-label="anchor" href="#the-genomicranges-package-and-classes"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="why-do-we-need-classes-for-genomic-ranges">Why do we need classes for genomic ranges?<a class="anchor" aria-label="anchor" href="#why-do-we-need-classes-for-genomic-ranges"></a></h3>
<p>In the era of genomics, many observations are reported as ranges of
coordinates - i.e., intervals - on a genomic scale. Depending on the
nature of the assay, those genomic ranges may represent genes,
transcripts, exons, single nucleotide polymorphisms (SNPs),
transcription factor binding sites, or peaks from next-generation
sequencing assays such as ChIP-seq or ATAC-seq.</p>
<p>Genomic ranges tie those observations of assayed values (e.g., gene
expression) to a physical location in the genome or an organism. For
instance, those genomic ranges can be used to query physical proximity
or overlap between assayed features and databases of known regulatory
regions.</p>
<p>Often, the final genomic ranges used for reporting measurements are
the result of combinations and operations on sets of genomic ranges in
databases of known genomic features. For instance, in RNA-sequencing,
next-generation sequencing reads are often counted within individual
exons, and those counts are subsequently aggregated across all the exons
of each gene. Separately, promoters are frequently defined as region of
arbitrary width, partly upstream and/or downstream of known
transcription start sites (TSS).</p>
<p>Importantly, genomic ranges do not necessarily need to span multiple
coordinates. The notion of range is meant in the mathematical way, and
single-nucleotide genomic ranges (e.g., SNPs) can be described as
opening and closing at the same coordinate (or at the next coordinate,
in the case of a right-open interval).</p>
<p>For many organisms, the genetic material is split into a number of
separate nucleic acid molecules (e.g., chromosomes, plasmids). As such,
genomic ranges are described by the name of the sequence and the numeric
interval of coordinates on that sequence.</p>
<figure><img src="../fig/bioc-genomicranges.svg" alt="Example uses of the GenomicRanges algebra." class="figure mx-auto d-block"></figure><p><strong>Example uses of the GenomicRanges algebra.</strong> Adapted
from Huber, Carey, Gentleman, Anders, Carlson, Carvalho, Bravo, Davis,
Gatto, Girke, Gottardo, Hahne, Hansen, Irizarry, Lawrence, Love,
MacDonald, Obenchain, Oles, Pages, Reyes, Shannon, Smyth, Tenenbaum,
Waldron, and Morgan (2015). The figure illustrates the example of a gene
model that comprises two transcripts, and the definition of various
genomic ranges relative to that gene model. For instance - in this
specific illustration - unspliced transcripts summarise the entire range
of coordinates from the start of the first exon to the end of the last
exon; while the gene region is defined as the set of coordinates
included in at least one exon of one transcript.</p>
</div>
<div class="section level3">
<h3 id="a-brief-introduction-to-intervals">A brief introduction to intervals<a class="anchor" aria-label="anchor" href="#a-brief-introduction-to-intervals"></a></h3>
<p>Intervals are described in mathematical terms using a start and an
end position on an axis of continuous coordinates. The interval then
comprises all the real numbers between those two coordinates, and the
width of each interval can be computed from the difference between the
coordinates of the start and end positions.</p>
<p>Generally speaking, the start and end position can be any rational
number, including floating-point numbers. However, in genomics, integer
coordinates are typically used to represent the location of monomers
(e.g., nucleotide, amino acid) in the sequence of polymers (e.g.,
nucleic acid, protein).</p>
<p>You may come across packages, databases, and programming languages
that use different rules to define intervals. In R, indexing is 1-based
(meaning that the first position in a sequence is 1), which contrasts
with Python that is 0-based (the index of the first position in a
sequence is 0). Similarly, references files in the UCSC Genome Browser
are 0-based, while those of the Ensembl Genome Browser are 1-based.</p>
<p>The definition of intervals in a shared coordinate system allows
calculations such as the distance between two intervals - generally
calculated as the distance between the two closest edges of those
intervals -, and the identification of overlapping intervals.</p>
<figure><img src="../fig/intervals.svg" alt="Example of intervals." class="figure mx-auto d-block"></figure><p><strong>Example of intervals.</strong> Three intervals named A, B,
and C, are represented. Interval A starts at position 5 and ends at
position 9, for a width of 4 units; interval B starts at position 1 and
ends at position 3, for a width of 2 units; interval C starts at
position 3 and ends at position 6, for a width of 3 units. Intervals A
and C overlap, from coordinates 5 to 6; while intervals B and C meet at
coordinate 3, but do not strictly overlap each other.</p>
</div>
<div class="section level3">
<h3 id="a-brief-introduction-to-genomic-ranges">A brief introduction to genomic ranges<a class="anchor" aria-label="anchor" href="#a-brief-introduction-to-genomic-ranges"></a></h3>
<p>Genomic ranges essentially extend the notion of mathematical
intervals on sets of biological sequences (e.g., chromosomes). In other
words, genomic ranges combine the name of the biological sequence on
which they are located with the integer range of coordinates that the
genomic ranges span in that sequence. This is key to distinguish genomic
features that span an overlapping range of coordinates on different
biological sequences.</p>
<p>Furthermore, the double-stranded nature of DNA sequences also adds
the notion of strandedness to genomic ranges. If known, the strand
information of genomic features is a key piece of information that
should be tracked, so that it may be used for downstream analyses. For
instance, genomic ranges spanning a common range of coordinates on
opposite strands of the same DNA sequence may not be considered to
overlap (e.g., for the purpose of strand-specific next-generation
sequencing assays).</p>
<p>Genomic ranges are <em>closed</em> intervals - the start and end
positions are included in the interval; in the example of nucleic acids,
the start position indicates the first nucleotide in the interval, and
the end position indicates the last nucleotide in the interval.</p>
<figure><img src="../fig/genomic-intervals.svg" alt="Example of genomic intervals." class="figure mx-auto d-block"></figure><p><strong>Example of genomic ranges.</strong> Genomic ranges are
defined by the name of the biological sequence in which they are located
(here, “chr1”), and the positions of start and end in that sequence.
Here, numeric positions are not explicitly shown, but implied by the
sequence of nucleic acids and the arrow indicating coordinates
increasing from the left to the right. In this example, genomic ranges
can be used to describe individual exons, with metadata grouping those
exons into transcripts and genes. Furthermore, the strandedness of
exons, transcripts, and genes is an important piece of information to
precisely describe the location of each genomic range in the
double-stranded DNA polymer.</p>
</div>
</section><section><h2 class="section-heading" id="the-genomicranges-package">The GenomicRanges package<a class="anchor" aria-label="anchor" href="#the-genomicranges-package"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a></h3>
<p>The <em><a href="https://bioconductor.org/packages/3.19/GenomicRanges" class="external-link">GenomicRanges</a></em>
package implements <a href="reference.html#s4-class">S4 classes</a> to
represent genomic ranges as S4 objects.</p>
<p>Specifically, the <code>GRanges</code> class is designed to store a
set of intervals including the name of the sequence where features are
located as well as the range of integer coordinates spanned by the
feature in that sequence.</p>
<p>More generally, the <code>IRanges</code> class is designed to store a
set of intervals over a range of integer coordinates, without the notion
of sequence names. As such, a <code>GRanges</code> object is merely the
combination of an <code>IRanges</code> object and a vector of sequence
names.</p>
<p>Those S4 classes provide automatic validity-checking functionality,
and a range of methods implementing common operations on integer
intervals and genomic ranges, from the calculation of distance between
pairs of intervals to the identification of overlapping genomic
ranges.</p>
<p>A short presentation of the basic classes defined in the <em><a href="https://bioconductor.org/packages/3.19/GenomicRanges" class="external-link">GenomicRanges</a></em>
package is available in one of the package vignettes, accessible as
<code>vignette("GenomicRangesIntroduction")</code>, while more detailed
information is provided in the other package vignettes, accessible as
<code>browseVignettes("GenomicRanges")</code>.</p>
</div>
<div class="section level3">
<h3 id="first-steps">First steps<a class="anchor" aria-label="anchor" href="#first-steps"></a></h3>
<p>To get started, we load the package.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">GenomicRanges</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="the-iranges-class">The IRanges class<a class="anchor" aria-label="anchor" href="#the-iranges-class"></a></h3>
<p>While the genomic space of many organisms is subdivided into multiple
sequences (e.g., chromosomes), many operations on genomic ranges take
place within individual sequences, where only integer positions matter.
The <code>IRanges</code> class provides a container for such “simple”
ranges that are defined by two out of three pieces of information:</p>
<ul><li>the start position of the range</li>
<li>the width of the range</li>
<li>the end position of the range</li>
</ul><p>The <code>IRanges()</code> constructor function accepts those three
pieces of information in the arguments <code>start=</code>,
<code>width=</code>, and <code>end=</code>. For instance, we create two
integer ranges from their start position and width:</p>
<ul><li>one range starts at position 10 and has width 10</li>
<li>one range starts at position 15 and has width 5</li>
</ul><div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_iranges</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">15</span><span class="op">)</span>, width <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">demo_iranges</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 2 ranges and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
  [1]        10        19        10
  [2]        15        19         5</code></pre>
</div>
<p>We note how the object displays not only the <em>start</em> and
<em>width</em> information that we requested for each range, but also
the <em>end</em> position that is naturally computed from the other two
pieces of information.</p>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Create the same two ranges as above, using the arguments
<code>start=</code> and <code>end=</code> of the <code>IRanges()</code>
constructor function.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">15</span><span class="op">)</span>, end <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">19</span>, <span class="fl">19</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 2 ranges and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
  [1]        10        19        10
  [2]        15        19         5</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The start and end positions as well as the width of every interval
can be extracted as numeric vector using the functions
<code>start()</code>, <code>end()</code> and <code>width()</code>,
respectively.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">start</span><span class="op">(</span><span class="va">demo_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 10 15</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">end</span><span class="op">(</span><span class="va">demo_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 19 19</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">width</span><span class="op">(</span><span class="va">demo_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 10  5</code></pre>
</div>
<p>Objects of the <code>IRanges</code> family extend the
<code>Vector</code> class, and are handled as unidimensional vectors in
terms of indexing. As such, individual ranges can be extracted by
integer index like any regular vector.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_iranges</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 1 range and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
  [1]        10        19        10</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="metadata-on-iranges">Metadata on IRanges<a class="anchor" aria-label="anchor" href="#metadata-on-iranges"></a></h3>
<p>The <code>IRanges</code> class can accommodate metadata information
on each range, including names - passed to the <code>names=</code>
argument - and miscellaneous metadata passed as named vectors.</p>
<p>For instance, we create two ranges named “A” and “B”. Furthermore, we
define metadata fields to store an example of character values and
numeric values, respectively. Both the names and the values of the
metadata fields are completely arbitrary in this example.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_with_metadata</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>  start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>,  <span class="fl">15</span><span class="op">)</span>,</span>
<span>  end   <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">19</span>,  <span class="fl">19</span><span class="op">)</span>,</span>
<span>  names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>  character_metadata <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"target"</span><span class="op">)</span>,</span>
<span>  numeric_metadata <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">100</span>, <span class="fl">200</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_with_metadata</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 2 ranges and 2 metadata columns:
        start       end     width | character_metadata numeric_metadata
    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |        &lt;character&gt;        &lt;numeric&gt;
  A        10        19        10 |            control              100
  B        15        19         5 |             target              200</code></pre>
</div>
<p>The metadata columns can be extracted as a <code>DataFrame</code>
using the function <code>mcols()</code> (short for “metadata
columns”).</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">demo_with_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 2 rows and 2 columns
  character_metadata numeric_metadata
         &lt;character&gt;        &lt;numeric&gt;
A            control              100
B             target              200</code></pre>
</div>
<p>The character vector of names can be extracted using the function
<code>names()</code>.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">names</span><span class="op">(</span><span class="va">demo_with_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A" "B"</code></pre>
</div>
<p>Similarly to named vector of base data types, individual ranges can
be extracted by name.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_with_metadata</span><span class="op">[</span><span class="st">"A"</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 1 range and 2 metadata columns:
        start       end     width | character_metadata numeric_metadata
    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |        &lt;character&gt;        &lt;numeric&gt;
  A        10        19        10 |            control              100</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="operations-on-iranges">Operations on IRanges<a class="anchor" aria-label="anchor" href="#operations-on-iranges"></a></h3>
<p><code>IRanges</code> provide the basis for most operations on ranges
of numerical coordinates.</p>
<p>For instance, given two sets of ranges - a query set and a subject
set - the <code>findOVerlaps()</code> function can be used to find out
which pairs of ranges in the two sets overlap with each other.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_iranges</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>  start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">8</span>, <span class="fl">16</span><span class="op">)</span>,</span>
<span>  end   <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">14</span>, <span class="fl">18</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">overlaps_iranges</span> <span class="op">&lt;-</span> <span class="fu">findOverlaps</span><span class="op">(</span>query <span class="op">=</span> <span class="va">query_iranges</span>, subject <span class="op">=</span> <span class="va">demo_iranges</span><span class="op">)</span></span>
<span><span class="va">overlaps_iranges</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Hits object with 3 hits and 0 metadata columns:
      queryHits subjectHits
      &lt;integer&gt;   &lt;integer&gt;
  [1]         1           1
  [2]         2           1
  [3]         2           2
  -------
  queryLength: 2 / subjectLength: 2</code></pre>
</div>
<p>The results are returned in the form of a <code>Hits</code> object,
which we have not introduced yet. A <code>Hits</code> object is
visualised as a table that comprises two integer columns named
<code>queryHits</code> and <code>subjectHits</code>. Each row in the
table reports an overlap between one range in the query set and one
range in the subject set, with the integer value in each column
indicating the index of the range in each set involved in the
overlap.</p>
<p>In this example, we confirm that the first range in the query set
overlaps the first range in the subject set; while the second range in
the query set overlaps both ranges in the subject set.</p>
<div id="going-further" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="going-further" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>For downstream use, the two components can be extracted from
<code>Hits</code> objects using their names, respectively:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">queryHits</span><span class="op">(</span><span class="va">overlaps_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 2</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subjectHits</span><span class="op">(</span><span class="va">overlaps_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 1 2</code></pre>
</div>
<p>While displayed as a table, <code>Hits</code> objects are actually
handled like vectors. Individual hits between one query range and one
subject range can be extracted their index:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">overlaps_iranges</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Hits object with 1 hit and 0 metadata columns:
      queryHits subjectHits
      &lt;integer&gt;   &lt;integer&gt;
  [1]         1           1
  -------
  queryLength: 2 / subjectLength: 2</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="the-granges-class">The GRanges class<a class="anchor" aria-label="anchor" href="#the-granges-class"></a></h3>
<p>Having defined integer ranges, the only additional information
necessary to define genomic ranges is the name of the genomic sequence
on which each range is located.</p>
<p>For instance, we define two genomic ranges, as follows:</p>
<ul><li>one genomic range on chromosome 1 (abbreviated “chr1”), from
position 10 to 25</li>
<li>one genomic range on chromosome 2 (abbreviated “chr2”), from
position 20 to 35</li>
</ul><p>To do so, we use the <code>GRanges()</code> constructor function. We
provide the sequence names as a character vector to the argument
<code>seqnames=</code>, and we provide both the start and end position
to the argument <code>ranges=</code> as an <code>IRanges</code>
object.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_granges</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>    start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    end   <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">25</span>, <span class="fl">35</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_granges</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1     10-25      *
  [2]     chr2     20-35      *
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<p>In the console, the object displays the sequence names in the
<code>seqnames</code> component, and the ranges in the form
<code>start-end</code> in the <code>ranges</code> component.
Furthermore, the example above also demonstrate that
<code>GRanges</code> objects possess a component called
<code>strand</code>; the symbol <code>*</code> indicates unstranded
genomic ranges, as we have not provided that information.</p>
<p>The strand information can be supplied to the <code>strand=</code>
argument, for instance:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_granges2</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>    start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    end   <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">25</span>, <span class="fl">35</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  strand  <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_granges2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1     10-25      +
  [2]     chr2     20-35      -
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<p>Finally, the example above also demonstrate that <code>GRanges</code>
objects include a component called <code>seqinfo</code>, which can be
used to store information about each sequence that may be represented in
the <code>seqnames</code> component. In the latest example above, we
have not provide any information about any sequence. As such, the
<code>seqinfo</code> component was automatically populated with the
names of the sequences that we used to create the object, while the
remaining pieces of information were left unspecified, as
<code>NA</code>.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqinfo</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Seqinfo object with 2 sequences from an unspecified genome; no seqlengths:
  seqnames seqlengths isCircular genome
  chr1             NA         NA   &lt;NA&gt;
  chr2             NA         NA   &lt;NA&gt;</code></pre>
</div>
<p>The example above reveals that information about sequences include
not only their respective name and length, but also whether they
represent a circular polymer (e.g., plasmid), and the name of the genome
that they are part of.</p>
<p>This information can be provided directly to the constructor when the
object is created, or edited on an existing object using the
<code>seqinfo()</code> accessor and the <code>Seqinfo()</code>
constructor:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqinfo</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span> <span class="op">&lt;-</span>  <span class="fu">Seqinfo</span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>    seqlengths <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1234</span>, <span class="fl">5678</span><span class="op">)</span>,</span>
<span>    isCircular <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    genome <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"homo_sapiens"</span>, <span class="st">"homo_sapiens"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_granges2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1     10-25      +
  [2]     chr2     20-35      -
  -------
  seqinfo: 2 sequences (1 circular) from homo_sapiens genome</code></pre>
</div>
<p>The start and end positions of the individual ranges as well as the
width of every interval can be extracted as numeric vector using the
functions <code>start()</code>, <code>end()</code> and
<code>width()</code>, respectively.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">start</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 10 20</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">end</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 25 35</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">width</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 16 16</code></pre>
</div>
<p>The sequence names and strand information can be extracted using the
functions <code>seqnames()</code> and <code>strand()</code>,
respectively.</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqnames</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>factor-Rle of length 2 with 2 runs
  Lengths:    1    1
  Values : chr1 chr2
Levels(2): chr1 chr2</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">strand</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>factor-Rle of length 2 with 2 runs
  Lengths: 1 1
  Values : + -
Levels(3): + - *</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="metadata-on-granges">Metadata on GRanges<a class="anchor" aria-label="anchor" href="#metadata-on-granges"></a></h3>
<p>Similarly to <code>IRanges</code>, metadata can be passed directly to
the <code>GRanges</code> constructor function. For instance:</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_granges3</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>    start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    end <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">25</span>, <span class="fl">35</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  metadata1 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"target"</span><span class="op">)</span>,</span>
<span>  metadata2 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_granges3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 2 ranges and 2 metadata columns:
      seqnames    ranges strand |   metadata1 metadata2
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
  [1]     chr1     10-25      * |     control         1
  [2]     chr2     20-35      * |      target         2
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="importing-genomic-ranges-from-files">Importing genomic ranges from files<a class="anchor" aria-label="anchor" href="#importing-genomic-ranges-from-files"></a></h3>
<p>Frequently, large collections of genomic ranges are imported from
files rather than described in manually written code. In particular,
genome-wide annotations of known gene features are distributed as files
on websites such as the <a href="https://www.ensembl.org/info/data/ftp/" class="external-link">Ensembl FTP</a> and the <a href="https://hgdownload.soe.ucsc.edu/downloads.html" class="external-link">UCSC Genome
Data</a> sites.</p>
<p>Various file formats are commonly used to store genomic ranges in
bioinformatics workflows. For instance, the BED (Browser Extensible
Data) format is commonly found in Chromatin Immunoprecipitation
Sequencing (ChIP-Seq), while GTF (Gene Transfer Format, GTF2.2) is the
<em>de facto</em> standard file format to describe genomic features such
as exons, transcripts, and genes.</p>
<p>In the following example, we import the gene model for Actin Beta
(ACTB) from a small GTF file as a set of genomic ranges. The example
file represents a subset of a GTF file for the <em>Homo sapiens</em>
species, downloaded from the <a href="https://www.ensembl.org/info/data/ftp/" class="external-link">Ensembl FTP</a> site. The
original file contains more than 3 millions lines and 22 metadata
fields, from which a subset was extracted into a smaller file for this
lesson.</p>
<p>In particular, we use the <code>import()</code> generic defined in
the <em><a href="https://bioconductor.org/packages/3.19/BiocIO" class="external-link">BiocIO</a></em>
package - with methods implemented in the <em><a href="https://bioconductor.org/packages/3.19/rtracklayer" class="external-link">rtracklayer</a></em>
package - as a versatile function that is capable of recognising common
file extensions and associating them with the appropriate method for
parsing each particular file format.</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">actb_gtf_data</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="st">"data/actb.gtf"</span><span class="op">)</span></span>
<span><span class="va">actb_gtf_data</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 267 ranges and 7 metadata columns:
        seqnames          ranges strand |      source           type     score
           &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt;
    [1]        7 5526409-5563902      - | rtracklayer     gene              NA
    [2]        7 5526409-5530601      - | rtracklayer     transcript        NA
    [3]        7 5530542-5530601      - | rtracklayer     exon              NA
    [4]        7 5529535-5529684      - | rtracklayer     exon              NA
    [5]        7 5529535-5529657      - | rtracklayer     CDS               NA
    ...      ...             ...    ... .         ...            ...       ...
  [263]        7 5540676-5540771      - | rtracklayer five_prime_utr        NA
  [264]        7 5529658-5529663      - | rtracklayer five_prime_utr        NA
  [265]        7 5561852-5562716      - | rtracklayer transcript            NA
  [266]        7 5562390-5562716      - | rtracklayer exon                  NA
  [267]        7 5561852-5561949      - | rtracklayer exon                  NA
            phase         gene_id   gene_name   transcript_id
        &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
    [1]      &lt;NA&gt; ENSG00000075624        ACTB            &lt;NA&gt;
    [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [5]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    ...       ...             ...         ...             ...
  [263]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [264]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [265]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000646584
  [266]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000646584
  [267]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000646584
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="going-further-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="going-further-1" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>Individual methods for parsing specific file formats can be invoked
directly. For instance, in this case, the GTF file format being
identical to the GFF version 2 file format, we could have directly
invoked the function <code>rtracklayer::import.gff2()</code> with the
exact same effect.</p>
<p>Refer to the documentation of the <em><a href="https://bioconductor.org/packages/3.19/rtracklayer" class="external-link">rtracklayer</a></em>
package for the full list of methods available.</p>
</div>
</div>
</div>
<p>In the example above, the contents of the GTF file were imported into
a <code>GRanges</code> object. For each entry in the file, the sequence
name, start and end position, and strand information were used to
populate the dedicated components of the object, while all other pieces
of information are stored as separate columns of metadata.</p>
<p>From here on, this <code>GRanges</code> object can be manipulated
just like any of the other <code>GRanges</code> objects that we have
created earlier in this episode.</p>
</div>
<div class="section level3">
<h3 id="operations-on-granges-and-the-grangeslist-class">Operations on GRanges and the GRangesList class<a class="anchor" aria-label="anchor" href="#operations-on-granges-and-the-grangeslist-class"></a></h3>
<p>As we have demonstrated so far, <code>GRanges</code> objects can be
manually defined or imported from files. Those often represent genomic
regions of interest, and databases of known genomic features,
respectively. Either way, a number of operations are commonly applied to
<code>GRanges</code> objects throughout bioinformatics workflows.</p>
<div class="section level4">
<h4 id="subset">Subset<a class="anchor" aria-label="anchor" href="#subset"></a></h4>
<p>For instance, the <code>subset()</code> method is extremely
convenient to extract a set of genomic ranges matching a condition on
any component, including sequence name, start and end position, strand,
or any metadata field. In the example below, we extract all the records
of type <code>transcript</code> that start at position
<code>5527147</code>.</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subset</span><span class="op">(</span><span class="va">actb_gtf_data</span>, <span class="va">type</span> <span class="op">==</span> <span class="st">"transcript"</span> <span class="op">&amp;</span> <span class="va">start</span> <span class="op">==</span> <span class="fl">5527147</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 5 ranges and 7 metadata columns:
      seqnames          ranges strand |      source       type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt;
  [1]        7 5527147-5529949      - | rtracklayer transcript        NA
  [2]        7 5527147-5530581      - | rtracklayer transcript        NA
  [3]        7 5527147-5530604      - | rtracklayer transcript        NA
  [4]        7 5527147-5530604      - | rtracklayer transcript        NA
  [5]        7 5527147-5530604      - | rtracklayer transcript        NA
          phase         gene_id   gene_name   transcript_id
      &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000642480
  [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000676397
  [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000676319
  [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000676189
  [5]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000473257
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="split">Split<a class="anchor" aria-label="anchor" href="#split"></a></h4>
<p>Separately, the <code>split()</code> method is useful to divide a set
of genomic ranges initially stored in a single <code>GRanges</code>
object into groups that are stored in a named list of
<code>GRanges</code> objects. Conveniently, the <code>GRangesList</code>
class provides a container for efficiently displaying and processing
lists of <code>GRanges</code> objects.</p>
<p>In the example below, we first extract the subset of entries that
represent exons, before separating those exons by transcript identifier,
yielding the result as a <code>GRangesList</code> object.</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">actb_exons</span> <span class="op">&lt;-</span> <span class="fu">subset</span><span class="op">(</span><span class="va">actb_gtf_data</span>, <span class="va">type</span> <span class="op">==</span> <span class="st">"exon"</span><span class="op">)</span></span>
<span><span class="va">actb_exons_by_transcript</span> <span class="op">&lt;-</span> <span class="fu">split</span><span class="op">(</span><span class="va">actb_exons</span>, <span class="va">actb_exons</span><span class="op">$</span><span class="va">transcript_id</span><span class="op">)</span></span>
<span><span class="va">actb_exons_by_transcript</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRangesList object of length 23:
$ENST00000414620
GRanges object with 4 ranges and 7 metadata columns:
      seqnames          ranges strand |      source     type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
  [1]        7 5562574-5562790      - | rtracklayer     exon        NA
  [2]        7 5540676-5540771      - | rtracklayer     exon        NA
  [3]        7 5529535-5529663      - | rtracklayer     exon        NA
  [4]        7 5529282-5529400      - | rtracklayer     exon        NA
          phase         gene_id   gene_name   transcript_id
      &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths

$ENST00000417101
GRanges object with 3 ranges and 7 metadata columns:
      seqnames          ranges strand |      source     type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
  [1]        7 5529806-5529982      - | rtracklayer     exon        NA
  [2]        7 5529535-5529663      - | rtracklayer     exon        NA
  [3]        7 5529235-5529400      - | rtracklayer     exon        NA
          phase         gene_id   gene_name   transcript_id
      &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000417101
  [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000417101
  [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000417101
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths

$ENST00000425660
GRanges object with 7 ranges and 7 metadata columns:
      seqnames          ranges strand |      source     type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
  [1]        7 5530524-5530601      - | rtracklayer     exon        NA
  [2]        7 5529535-5529663      - | rtracklayer     exon        NA
  [3]        7 5529161-5529400      - | rtracklayer     exon        NA
  [4]        7 5529019-5529059      - | rtracklayer     exon        NA
  [5]        7 5528281-5528719      - | rtracklayer     exon        NA
  [6]        7 5528004-5528185      - | rtracklayer     exon        NA
  [7]        7 5527156-5527891      - | rtracklayer     exon        NA
          phase         gene_id   gene_name   transcript_id
      &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [5]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [6]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [7]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths

...
&lt;20 more elements&gt;</code></pre>
</div>
<p>When printing the object above in the console, the first line
confirms the class of the object as <code>GRrangesList</code>, while
each named <code>GRanges</code> in that list is introduced by the dollar
sign and the name of that item, just like regular named lists in base
R.</p>
</div>
<div class="section level4">
<h4 id="length">Length<a class="anchor" aria-label="anchor" href="#length"></a></h4>
<p>By nature, many of the methods applicable to <code>list</code>
objects can be directly applied to <code>GRangesList</code> objects. For
instance, the <code>lengths()</code> function can be used on
<code>GRangesList</code> to display the length of each
<code>GRanges</code> object in the list as an integer vector.</p>
<p>In the latest example above, we can compute the number of exons in
each transcript as the length of each <code>GRanges</code> object within
the <code>GRangesList</code>:</p>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">lengths</span><span class="op">(</span><span class="va">actb_exons_by_transcript</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ENST00000414620 ENST00000417101 ENST00000425660 ENST00000432588 ENST00000443528
              4               3               7               5               3
ENST00000462494 ENST00000464611 ENST00000473257 ENST00000477812 ENST00000480301
              5               3               5               5               2
ENST00000484841 ENST00000493945 ENST00000642480 ENST00000645025 ENST00000645576
              5               6               5               4               5
ENST00000646584 ENST00000646664 ENST00000647275 ENST00000674681 ENST00000675515
              2               6               3               6               6
ENST00000676189 ENST00000676319 ENST00000676397
              6               3               6 </code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Importantly, the function <code>lengths()</code> (with a final
<code>s</code>) demonstrated above is different from the function
<code>length()</code> (without <code>s</code>). The former is meant to
be used on list objects, returning a vector giving the length of each
element in the list; while the latter returns a single numeric scalar
giving the length of the list itself (i.e., the number of elements in
the list).</p>
<p>What does <code>length(actb_exons_by_transcript)</code> return, and
what does this number represent biologically?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">actb_exons_by_transcript</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23</code></pre>
</div>
<p>This code returns the single integer value <code>23</code>, which is
the number of <code>GRanges</code> in the <code>GRangesList</code>
object and the number of transcripts for the gene ACTB.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="subset-by-overlap">Subset by overlap<a class="anchor" aria-label="anchor" href="#subset-by-overlap"></a></h3>
<p>Possibly one of the most common operations when working with genomic
ranges is to subset arbitrarily large collections of genomic ranges to
those located in a specific region of the genome; for instance, when
visualising information as tracks in a genome browser.</p>
<p>To demonstrate, we manually define a new <code>GRanges</code>
representing a region of interest that we will use to extract all of the
genomic ranges imported earlier from the GTF file which overlap that
region of interest.</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_of_interest</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="st">"7"</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fl">5525830</span>, end <span class="op">=</span> <span class="fl">5531239</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">actb_in_region</span> <span class="op">&lt;-</span> <span class="fu">subsetByOverlaps</span><span class="op">(</span>x <span class="op">=</span> <span class="va">actb_gtf_data</span>, ranges <span class="op">=</span> <span class="va">region_of_interest</span><span class="op">)</span></span>
<span><span class="va">actb_in_region</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 256 ranges and 7 metadata columns:
        seqnames          ranges strand |      source           type     score
           &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt;
    [1]        7 5526409-5563902      - | rtracklayer     gene              NA
    [2]        7 5526409-5530601      - | rtracklayer     transcript        NA
    [3]        7 5530542-5530601      - | rtracklayer     exon              NA
    [4]        7 5529535-5529684      - | rtracklayer     exon              NA
    [5]        7 5529535-5529657      - | rtracklayer     CDS               NA
    ...      ...             ...    ... .         ...            ...       ...
  [252]        7 5529535-5529657      - | rtracklayer CDS                   NA
  [253]        7 5529655-5529657      - | rtracklayer start_codon           NA
  [254]        7 5529282-5529400      - | rtracklayer exon                  NA
  [255]        7 5529282-5529400      - | rtracklayer CDS                   NA
  [256]        7 5529658-5529663      - | rtracklayer five_prime_utr        NA
            phase         gene_id   gene_name   transcript_id
        &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
    [1]      &lt;NA&gt; ENSG00000075624        ACTB            &lt;NA&gt;
    [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [5]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    ...       ...             ...         ...             ...
  [252]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [253]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [254]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [255]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [256]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<p>Like the <code>subset()</code> method, the
<code>subsetByOverlaps()</code> method returns a new
<code>GRanges</code> object. We can visually compare the information
printed in the object (256 ranges in the new subsetted object, relative
to 267 ranges in the original object), or we can programmatically
compare the length of the two objects to check whether the new
<code>GRanges</code> object is any smaller than the original
<code>GRanges</code> object:</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">actb_in_region</span><span class="op">)</span> <span class="op">-</span> <span class="fu">length</span><span class="op">(</span><span class="va">actb_gtf_data</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] -11</code></pre>
</div>
<p>In the example above, we learn that the new <code>GRanges</code>
object has 11 records less than the original <code>GRanges</code>
object.</p>
<div id="going-further-2" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="going-further-2" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>Many more methods exist to operate on <code>GRanges</code> and
<code>GRangesList</code> objects than what could be demonstrated
here.</p>
<p>You can find the full list of functions defined in the
<code>GenomicRanges</code> package on the index page of the package
documentation, accessible using
<code><a href="https://rdrr.io/pkg/GenomicRanges/man" class="external-link">help(package="GenomicRanges")</a></code>. You can also find more
examples and use cases in the package vignettes, accessible using
<code>browseVignettes("GenomicRanges")</code>.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>The <code>GenomicRanges</code> package defines classes to represent
ranges of coordinates on a genomic scale.</li>
<li>The <code>GenomicRanges</code> package also defines methods to
efficiently process genomic ranges.</li>
<li>The <code>rtracklayer</code> package provides functions to import
and export genomic ranges from and to common genomic file formats.</li>
</ul></div>
</div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/06-biological-sequences.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/08-annotations.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/06-biological-sequences.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Working with
        </a>
        <a class="chapter-link float-end" href="../instructor/08-annotations.html" rel="next">
          Next: Working with...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/07-genomic-ranges.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-project/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-project/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-project/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/bioc-project/instructor/07-genomic-ranges.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Working with genomics ranges",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-project/instructor/07-genomic-ranges.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-project/instructor/07-genomic-ranges.html",
  "dateCreated": "2020-09-14",
  "dateModified": "2025-06-09",
  "datePublished": "2025-06-17"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

